pkg/runtime/scheduler/scheduler.go

```diff
@@ -15,90 +15,108 @@ package scheduler
 
-	"github.com/dapr/dapr/pkg/channel"
-	v1 "github.com/dapr/dapr/pkg/messaging/v1"
+	"github.com/dapr/dapr/pkg/actors"
+	invokev1 "github.com/dapr/dapr/pkg/messaging/v1"
+	internalv1pb "github.com/dapr/dapr/pkg/proto/internals/v1"
 	schedulerv1pb "github.com/dapr/dapr/pkg/proto/scheduler/v1"
+	"github.com/dapr/dapr/pkg/resiliency"
+	"github.com/dapr/dapr/pkg/runtime/channels"
+	"github.com/dapr/dapr/pkg/runtime/scheduler/clients"
 	"github.com/dapr/dapr/pkg/scheduler/client"
-	"github.com/dapr/dapr/pkg/security"
 	"github.com/dapr/kit/concurrency"
 	"github.com/dapr/kit/logger"
+	"github.com/golang/protobuf/ptypes/wrappers"
 )
 
 var log = logger.NewLogger("dapr.runtime.scheduler")
 
-// Manager manages connections to multiple schedulers.
-type Manager struct {
-	clients     []*client.Client
-	namespace   string
-	appID       string
-	lastUsedIdx int64
-	appChannel  channel.AppChannel
-	isHTTP      bool
+type Options struct {
+	Namespace string
+	AppID     string
+	Clients   *clients.Clients
+	Resilicy  resiliency.Provider
+	Channels  *channels.Channels
 }
 
-type Options struct {
-	Namespace  string
-	AppID      string
-	Addresses  string
-	Security   security.Handler
-	AppChannel channel.AppChannel
-	IsHTTP     bool
+// Manager manages connections to multiple schedulers.
+type Manager struct {
+	clients    *clients.Clients
+	namespace  string
+	appID      string
+	actors     actors.ActorRuntime
+	resiliency resiliency.Provider
+	channels   *channels.Channels
+
+	startCh chan struct{}
+	started atomic.Bool
+	wg      sync.WaitGroup
+	jobCh   chan *schedulerv1pb.WatchJobsResponse
+	closeCh chan struct{}
+	running atomic.Bool
 }
 
-func NewManager(ctx context.Context, opts Options) *Manager {
-	manager := &Manager{
+func New(opts Options) (*Manager, error) {
+	return &Manager{
 		namespace:  opts.Namespace,
 		appID:      opts.AppID,
-		appChannel: opts.AppChannel,
-		isHTTP:     opts.IsHTTP,
+		clients:    opts.Clients,
+		resiliency: opts.Resilicy,
+		channels:   opts.Channels,
+		jobCh:      make(chan *schedulerv1pb.WatchJobsResponse),
+		startCh:    make(chan struct{}),
+		closeCh:    make(chan struct{}),
+	}, nil
+}
+
+// Run starts watching for job triggers from all scheduler clients.
+func (m *Manager) Run(ctx context.Context) error {
+	if !m.running.CompareAndSwap(false, true) {
+		return errors.New("scheduler manager is already running")
 	}
 
-	schedulerHostPorts := strings.Split(opts.Addresses, ",")
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-m.startCh:
+	}
 
-	for _, address := range schedulerHostPorts {
-		log.Debugf("Attempting to connect to Scheduler at address: %s", address)
-		conn, cli, err := client.New(ctx, address, opts.Security)
-		if err != nil {
-			log.Debugf("Scheduler client not initialized for address: %s", address)
-		} else {
-			log.Debugf("Scheduler client initialized for address: %s", address)
+	clients := m.clients.All()
+	runners := make([]concurrency.Runner, len(clients), len(clients)+2)
```

Comment:
whats the +2 for? Should that be dynamic based on the len(schedulerAddresses)?

Reply:
For the two runners added below

```go
	runners = append(runners,
		m.processTriggers,
		func(ctx context.Context) error {
			<-ctx.Done()
			close(m.closeCh)
			return nil
		},
	)

```